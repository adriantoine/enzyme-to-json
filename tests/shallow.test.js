/* eslint-env jest */

import React from 'react';
import omitBy from 'lodash/omitBy';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import shallowToJson from '../src/shallow';
import {
  BasicPure,
  BasicWithUndefined,
  BasicWithAList,
  ArrayRender,
  FalsyTruthyComponent,
  FalsyChildren,
  FragmentAsChild,
  FragmentAsRoot,
  SuspenseAsChild,
  SuspenseAsRoot,
  ComponentWithMemo,
  WithDefaultProps,
  ComponentWithChildren,
} from './fixtures/pure-function';
import {
  BasicClass,
  ClassWithPure,
  ClassWithNull,
  ClassArrayRender,
  ClassWithDefaultProps,
} from './fixtures/class';

import {
  ForwardRefWithDefaultProps,
} from './fixtures/forwardRef';

Enzyme.configure({adapter: new Adapter()});

function WrapperComponent(props) {
  return <BasicPure {...props} />;
}

it('doesnt break when called without arguments', () => {
  expect(shallowToJson()).toBe(null);
  expect(shallowToJson([null])).toBe(null);
  expect(shallowToJson(['a'])).toBe(null);
});

it('converts basic pure shallow', () => {
  const shallowed = shallow(
    <BasicPure className="pure">
      <strong>Hello!</strong>
    </BasicPure>,
  );

  expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('converts basic class shallow', () => {
  const shallowed = shallow(
    <BasicClass className="class">
      <strong>Hello!</strong>
    </BasicClass>,
  );

  expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('converts a class mount with a pure function in it', () => {
  const shallowed = shallow(
    <ClassWithPure className="class">
      <strong>Hello!</strong>
    </ClassWithPure>,
  );

  expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('handles a component which returns null', () => {
  const shallowed = shallow(<ClassWithNull />);
  expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('handles elements in props', () => {
  const shallowed = shallow(
    <WrapperComponent
      element={
        <BasicPure>
          <strong>Hello!</strong>
        </BasicPure>
      }
    />,
  );
  expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('handles elements in prop arrays', () => {
  const shallowed = shallow(
    <WrapperComponent
      elements={[
        <BasicPure>
          <strong>Hello!</strong>
        </BasicPure>,
      ]}
    />,
  );

  expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('handles elements in prop objects', () => {
  const shallowed = shallow(
    <WrapperComponent
      element={{
        element: (
          <BasicPure>
            <strong>Hello!</strong>
          </BasicPure>
        ),
        nestedElements: [
          <BasicPure>
            <strong>Hello again!</strong>
          </BasicPure>,
        ],
      }}
    />,
  );

  expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('ignores non-plain objects', () => {
  function TestConstructor() {
    this._test = true;
  }

  const shallowed = shallow(
    <WrapperComponent instance={new TestConstructor()} />,
  );

  expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('skips undefined props', () => {
  const shallowed = shallow(<BasicWithUndefined>Hello!</BasicWithUndefined>);

  expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('renders multiple elements as a result of find', () => {
  const shallowed = shallow(<BasicWithAList />);
  expect(shallowToJson(shallowed.find('li'))).toMatchSnapshot();
});

it('outputs the key prop', () => {
  const shallowed = shallow(
    <ul>
      <li key={1} />
      <li key={2} />
    </ul>,
  );
  expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('doesnt output the key prop when noKey option is passed', () => {
  const shallowed = shallow(
    <ul>
      <li key={1} />
      <li key={2} />
    </ul>,
  );
  expect(shallowToJson(shallowed, {noKey: true})).toMatchSnapshot();
});

it('converts function components with render returning top level arrays', () => {
  const shallowed = shallow(
    <ArrayRender>
      <strong>Hello!</strong>
    </ArrayRender>,
  );

  expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('converts class components with render returning top level arrays', () => {
  const shallowed = shallow(
    <ClassArrayRender>
      <strong>Hello!</strong>
    </ClassArrayRender>,
  );

  expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('accepts a map option allowing to customize content', () => {
  const shallowed = shallow(<strong>Hello!</strong>);

  expect(
    shallowToJson(shallowed, {
      map: json => ({...json, children: ['Goodbye!']}),
    }),
  ).toMatchSnapshot();
});

it('accepts a map option allowing to customize content of all nested components', () => {
  const shallowed = shallow(
    <div randomlyGeneratedKey={Date.now()} className="wrapper">
      <strong randomlyGeneratedKey={Date.now()}>Hello!</strong>
      <strong className="strong2">Hello 2</strong>
    </div>,
  );

  expect(
    shallowToJson(shallowed, {
      map: json => ({
        ...json,
        props: omitBy(json.props, (val, key) => key === 'randomlyGeneratedKey'),
      }),
    }),
  ).toMatchSnapshot();
});

it('can skip a component I dont want to see with the map option', () => {
  const shallowed = shallow(
    <div>
      <ul>
        <li>Item 1</li>
        <li>Item 2</li>
      </ul>
      <strong>Hello 2</strong>
    </div>,
  );

  expect(
    shallowToJson(shallowed, {
      map: json => {
        if (json.type === 'ul') {
          return null;
        }
        return json;
      },
    }),
  ).toMatchSnapshot();
});

it('outputs the snapshot even with inline JSX conditions being falsy', () => {
  const shallowed = shallow(
    <div>
      <span>I am there</span>
      {false && <span>Issue</span>}
    </div>,
  );

  expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('outputs an empty string when a component returns false', () => {
  const shallowed = shallow(<FalsyTruthyComponent foo={false} />);
  expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('outputs an empty string when a component returns undefined', () => {
  const shallowed = shallow(<FalsyTruthyComponent />);
  expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('outputs an empty string when a component has false chidren', () => {
  const shallowed = shallow(<FalsyChildren falsy={false} />);
  expect(shallowToJson(shallowed)).toMatchSnapshot();
});

it('renders a component that has a child fragment', () => {
  const wrapper = shallow(<FragmentAsChild />);

  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

it('renders a component that has a fragment root', () => {
  const wrapper = shallow(<FragmentAsRoot />);

  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

it('renders a component that has a Suspense with Lazy as child', () => {
  const wrapper = shallow(<SuspenseAsChild />);

  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

it('renders a component that has a Suspense as root with Lazy child', () => {
  const wrapper = shallow(<SuspenseAsRoot />);

  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

it('renders a component that has a Memo component', () => {
  const wrapper = shallow(<ComponentWithMemo />);

  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

it('should bleed default props from functional child component into snapshot without the option', () => {
  const wrapper = shallow(
    <ComponentWithChildren>
      <WithDefaultProps />
    </ComponentWithChildren>,
  );

  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

it('should not bleed default props from functional child component into snapshot', () => {
  const wrapper = shallow(
    <ComponentWithChildren>
      <WithDefaultProps />
    </ComponentWithChildren>,
  );

  expect(shallowToJson(wrapper, { ignoreDefaultProps: true })).toMatchSnapshot();
});

it('should set prop that has a different value from default prop values of functional component', () => {
  const wrapper = shallow(
    <ComponentWithChildren>
      <WithDefaultProps value="yeah, man" />
    </ComponentWithChildren>,
  );

  expect(shallowToJson(wrapper, { ignoreDefaultProps: true })).toMatchSnapshot();
});

it('should bleed default props from class child component into snapshot without the option', () => {
  const wrapper = shallow(
    <ComponentWithChildren>
      <ClassWithDefaultProps />
    </ComponentWithChildren>,
  );

  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

it('should not bleed default props from class child component into snapshot', () => {
  const wrapper = shallow(
    <ComponentWithChildren>
      <ClassWithDefaultProps />
    </ComponentWithChildren>,
  );

  expect(shallowToJson(wrapper, { ignoreDefaultProps: true })).toMatchSnapshot();
});

it('should set prop that has a different value from default prop values of class component', () => {
  const wrapper = shallow(
    <ComponentWithChildren>
      <ClassWithDefaultProps value="ah, man" />
    </ComponentWithChildren>,
  );

  expect(shallowToJson(wrapper, { ignoreDefaultProps: true })).toMatchSnapshot();
});

it('should bleed default props from forwardRef child component into snapshot without the option', () => {
  const wrapper = shallow(
    <ComponentWithChildren>
      <ForwardRefWithDefaultProps />
    </ComponentWithChildren>,
  );

  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

it('should not bleed default props from forwardRef child component into snapshot', () => {
  const wrapper = shallow(
    <ComponentWithChildren>
      <ForwardRefWithDefaultProps />
    </ComponentWithChildren>,
  );

  expect(shallowToJson(wrapper, { ignoreDefaultProps: true })).toMatchSnapshot();
});

it('should set prop that has a different value from default prop values of forwardRef component', () => {
  const wrapper = shallow(
    <ComponentWithChildren>
      <ForwardRefWithDefaultProps value="yeah, man" />
    </ComponentWithChildren>,
  );

  expect(shallowToJson(wrapper, { ignoreDefaultProps: true })).toMatchSnapshot();
});
