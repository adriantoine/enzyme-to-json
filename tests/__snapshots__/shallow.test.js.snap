// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accepts a map option allowing to customize content 1`] = `
<strong>
  Goodbye!
</strong>
`;

exports[`accepts a map option allowing to customize content of all nested components 1`] = `
<div
  className="wrapper"
>
  <strong>
    Hello!
  </strong>
  <strong
    className="strong2"
  >
    Hello 2
  </strong>
</div>
`;

exports[`can skip a component I dont want to see with the map option 1`] = `
<div>
  <strong>
    Hello 2
  </strong>
</div>
`;

exports[`converts a class mount with a pure function in it 1`] = `
<div
  className="class-with-pure class"
  onClick={[Function]}
>
  <BasicPure
    className="nested-pure"
  >
    <span>
      <strong>
        Hello!
      </strong>
    </span>
    <span
      className="empty"
    />
  </BasicPure>
</div>
`;

exports[`converts basic class shallow 1`] = `
<div
  className="basic-class class"
  onClick={[Function]}
>
  <div
    className="group"
    id="group-id"
  >
    <span>
      <strong>
        Hello!
      </strong>
    </span>
    <span
      className="empty"
    />
  </div>
</div>
`;

exports[`converts basic pure shallow 1`] = `
<div
  className="basic-pure pure"
  onClick={[Function]}
>
  <div
    className="group"
    id="group-id"
  >
    <span>
      <strong>
        Hello!
      </strong>
    </span>
    <span
      className="empty"
    />
  </div>
</div>
`;

exports[`converts class components with render returning top level arrays 1`] = `
Array [
  <div
    className="test"
    key="test"
  >
    Test
  </div>,
  <div
    className="test2"
    key="test2"
  >
    Test 2
  </div>,
  <div
    className="child"
    key="child"
  >
    <strong>
      Hello!
    </strong>
  </div>,
]
`;

exports[`converts function components with render returning top level arrays 1`] = `
Array [
  <div
    className="test"
    key="test"
  >
    Test
  </div>,
  <div
    className="test2"
    key="test2"
  >
    Test 2
  </div>,
  <div
    className="child"
    key="child"
  >
    <strong>
      Hello!
    </strong>
  </div>,
]
`;

exports[`doesnt output the key prop when noKey option is passed 1`] = `
<ul>
  <li />
  <li />
</ul>
`;

exports[`handles a component which returns null 1`] = `""`;

exports[`handles elements in prop arrays 1`] = `
<BasicPure
  elements={
    Array [
      <BasicPure>
        <strong>
          Hello!
        </strong>
      </BasicPure>,
    ]
  }
/>
`;

exports[`handles elements in prop objects 1`] = `
<BasicPure
  element={
    Object {
      "element": <BasicPure>
        <strong>
          Hello!
        </strong>
      </BasicPure>,
      "nestedElements": Array [
        <BasicPure>
          <strong>
            Hello again!
          </strong>
        </BasicPure>,
      ],
    }
  }
/>
`;

exports[`handles elements in props 1`] = `
<BasicPure
  element={
    <BasicPure>
      <strong>
        Hello!
      </strong>
    </BasicPure>
  }
/>
`;

exports[`ignores non-plain objects 1`] = `
<BasicPure
  instance={
    TestConstructor {
      "_test": true,
    }
  }
/>
`;

exports[`outputs an empty string when a component has false chidren 1`] = `<div />`;

exports[`outputs an empty string when a component returns false 1`] = `""`;

exports[`outputs an empty string when a component returns undefined 1`] = `""`;

exports[`outputs the key prop 1`] = `
<ul>
  <li
    key="1"
  />
  <li
    key="2"
  />
</ul>
`;

exports[`outputs the snapshot even with inline JSX conditions being falsy 1`] = `
<div>
  <span>
    I am there
  </span>
</div>
`;

exports[`renders a component that has a Memo component 1`] = `
<div>
  <span />
  <Memo() />
</div>
`;

exports[`renders a component that has a Suspense as root with Lazy child 1`] = `
<Suspense>
  <span />
  <div />
  <lazy />
</Suspense>
`;

exports[`renders a component that has a Suspense with Lazy as child 1`] = `
<div>
  <Suspense>
    <lazy />
    <div />
    <button />
  </Suspense>
</div>
`;

exports[`renders a component that has a child fragment 1`] = `
<div>
  <span />
  <div />
  <button />
</div>
`;

exports[`renders a component that has a fragment root 1`] = `
<Fragment>
  <span />
  <div />
  <button />
</Fragment>
`;

exports[`renders multiple elements as a result of find 1`] = `
Array [
  <li>
    0
  </li>,
  <li>
    1
  </li>,
]
`;

exports[`should bleed default props from class child component into snapshot without the option 1`] = `
<span>
  <ClassWithDefaultProps
    falsyValue={false}
    value="hi mum"
  />
</span>
`;

exports[`should bleed default props from forwardRef child component into snapshot without the option 1`] = `
<span>
  <ForwardRef
    falsyValue={false}
    value="hi mum"
  />
</span>
`;

exports[`should bleed default props from functional child component into snapshot without the option 1`] = `
<span>
  <WithDefaultProps
    falsyValue={false}
    value="hi there"
  />
</span>
`;

exports[`should not bleed default props from class child component into snapshot 1`] = `
<span>
  <ClassWithDefaultProps />
</span>
`;

exports[`should not bleed default props from forwardRef child component into snapshot 1`] = `
<span>
  <ForwardRef />
</span>
`;

exports[`should not bleed default props from functional child component into snapshot 1`] = `
<span>
  <WithDefaultProps />
</span>
`;

exports[`should set prop that has a different value from default prop values of class component 1`] = `
<span>
  <ClassWithDefaultProps
    value="ah, man"
  />
</span>
`;

exports[`should set prop that has a different value from default prop values of forwardRef component 1`] = `
<span>
  <ForwardRef
    value="yeah, man"
  />
</span>
`;

exports[`should set prop that has a different value from default prop values of functional component 1`] = `
<span>
  <WithDefaultProps
    value="yeah, man"
  />
</span>
`;

exports[`skips undefined props 1`] = `
<button>
  Hello
</button>
`;
